### PR TITLE
Compiler specification

### PR DESCRIPTION
**Pull request by gabrielhdt**
Interface for compiler and some tests on compiler pure interface

### CODE REVIEW

File: src/handle/compile.mli [hunk 1]
```diff
@@ -0,0 +1,33 @@
+(** High-level compilation functions. *)
+
+open Common
+open Core
+
+val gen_obj : bool Stdlib.ref
+(** [gen_obj] indicates whether we should generate object files when compiling
+    source files. The default behaviour is not te generate them. *)
```
Comment (fblanqui) [inline src/handle/compile.mli:8]: small typo: not te -> to not

File: src/handle/compile.mli [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+val compile_file : string -> Sign.t
+(** [compile_file fname] looks for a package configuration file for
+    [fname] and compiles [fname]. It is the main compiling function. It
+    is called from the main program exclusively. *)
+
+open Console
+
+(** Pure wrappers around compilation functions. Functions provided perform the
+    same computations as the ones defined earlier, but restores the state when
```
Comment (fblanqui) [inline src/handle/compile.mli:27]: typos: Functions provided -> The provided functions, earlier -> above, restores -> restore

Comment (fblanqui) [reply src/handle/compile.mli:27]: What about renaming compile_file into read_pkg_file_and_compile?

**Comment by fblanqui** [comment]
Is this PR related to #403?

File: src/handle/compile.mli [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+open Core
+
+val gen_obj : bool Stdlib.ref
+(** [gen_obj] indicates whether we should generate object files when compiling
+    source files. The default behaviour is not te generate them. *)
+
+val recompile : bool Stdlib.ref
+(** [recompile] indicates whether we should recompile files who have an object
+    file that is already up to date. Note that this flag only applies to files
+    that are given on the command line explicitly, not their dependencies. *)
```
[the remark this answers is not in the document]
Comment (gabrielhdt) [reply src/handle/compile.mli:13]: I don't have a clear opinion on this. It's related to `--recompile`, so it impacts user quite directly. I don't know if anyone uses this option.

**Comment by gabrielhdt** [comment]
> Is this PR related to #403?

No it's not. We just introduce some new tests on the compile functions.

**Comment by gabrielhdt** [comment]
For the links between `Compile.Pure` and `Pure`, see https://github.com/Deducteam/lambdapi/pull/457#issuecomment-724752906

**Comment by fblanqui** [comment]
@gabrielhdt Do you want to merge it now?

**Pull request merged**